### PR TITLE
Document IPNI HTTP provider specification

### LINKED ISSUE #16: Formalize ipni advertisement publishing http spec
currently this is the behavior described in https://github.com/ipni/specs/blob/main/IPNI.md#http-1

we probably want to express a specific implicit path we might want as a root for the advertisement chain over http.

This would be a more specific doc targeted at defining what a provider would need to do to be considered an 'http index provider'

### PR DESCRIPTION
**Pull request by masih**
[Rendered](https://github.com/ipni/specs/blob/masih/http_prov_spec/IPNI_HTTP_PROVIDER.md)

**Note: I have opted for least disruptive spec relative to the current implementation. If you folks think this doesn't matter as much there are things that I'd like to change.**

Fixes #16

### CODE REVIEW

**Review by willscott — APPROVED** [review]
this looks like a good extraction of what we expect from the provider

File: IPNI_HTTP_PROVIDER.md
Comment (rvagg) [inline IPNI_HTTP_PROVIDER.md:234]:
```suggestion
        "bytes": "EiDYa/LrNYDdHjSlKGXpHjc1IyHds9Xdi/p8d25Q5UE/bQ"
```

(these padded examples need to disappear into history)

**Review by gammazero — APPROVED** [review]
This is excellent

**Comment by rvagg** [comment]
It might be nice to leave open (and document here) the ability of announcement `Addrs` to include optional [`httpath`](https://github.com/ipni/go-libipni/blob/1bd7b3f90d2a849c3449742947e86d78e66508ab/maurl/convert.go#L35-L41) specifier, and be clear how this interacts with the new path prefixes defined here. If I provide an `httpath` in a multiaddr, does it override `/ipni/v1/ad/`, or does it prepend to this?
My own limited use of this has involved wanting to change the default provider server that responds to `/head` and `/{cid}` to one that responds on `/_ipni/head` and `/_ipni/{cid}` because I want to [reuse an existing server that has other things going on at the root](https://github.com/ipni/go-libipni/pull/34). So I like that this is now namespaced; I also like optionality, but it's not critical now that this is so specific. i.e. custom `httpath` could be ruled out in this spec and I wouldn't be too upset, but I'd like the clarity.

File: IPNI_HTTP_PROVIDER.md
Comment (ischasny) [inline IPNI_HTTP_PROVIDER.md:184]: Would it be helpful to have bulk fetches in the specification? Something like `GET /ipni/v1/ad/{CID}?depth=20`, so that chains can be traversed faster. Support for depth can be optional too.

Comment (masih) [reply IPNI_HTTP_PROVIDER.md:184]: Out of scope for this spec IMO. 

How it occurs for me is that future iterations can introduce this sort of functionality, along with support for `[application/vnd.ipld.car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car)` where we can ask a provider to preassemble CAR files containing ads.

This functionality is sufficiently complex that deserves its own spec/iteration IMHO. 

Cc @gammazero

Comment (ischasny) [reply IPNI_HTTP_PROVIDER.md:184]: Sounds good, this is more of a optimisation rather than a must have. Other than that - looks good to me.

**Review by ischasny — APPROVED** [review]

**Review by ischasny — APPROVED** [review]

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:206]: Added this as an item in [future considerations](https://github.com/ipni/specs/blob/masih/http_prov_spec/IPNI_HTTP_PROVIDER.md#future-considerations)

**Comment by rvagg** [comment]
great! I've followed up in https://github.com/ipni/go-libipni/issues/42

**Comment by masih** [comment]
Cc @lidel @aschmahmann

**Comment by BigLep** [comment]
This came up in https://pl-strflt.notion.site/2023-05-30-Content-Routing-WG-12-b2ed74834fe44e359bbcdd02740e2084

There is going to be implementation need for this in the next quarter or two.  As a result, want to get ahead before code gets written and decisions ossify.

Some next steps:
1. Update the IPIP to the newest template standard with lessons learned the last 6 months
2. Review from spec maintainers

Implementation notes:
"Lassie currently does not depend on boxo. It may be hard for us to prototype ambient-discovery in boxo until the boxo/rest-of-world dependency conflicts are resolved"

(I'll come through and update this more later today.)

File: IPNI_HTTP_PROVIDER.md
Comment (BigLep) [inline IPNI_HTTP_PROVIDER.md:70]:
```suggestion
   conflict with other APIs such as the [IPFS HTTP Gateway](https://specs.ipfs.tech/http-gateways/) or Boost.
```

File: IPNI_HTTP_PROVIDER.md
Comment (BigLep) [inline IPNI_HTTP_PROVIDER.md:84]: Your call, but if you're expecting others outside PL to read this, could be good to link to sources for more information.

Maybe use reference style links?

File: IPNI_HTTP_PROVIDER.md
Comment (BigLep) [inline IPNI_HTTP_PROVIDER.md:143]: 1. (newbie question) Is `httpath` something we're coming up with now, or is there existing discussion on this?
2. Is this intentional to be `httpath` rather than `httppath` ?

File: IPNI_HTTP_PROVIDER.md
Comment (BigLep) [inline IPNI_HTTP_PROVIDER.md:272]: Don't you want to standardize how paging is communicated between clients and servers to avoid divergence between implementations?

File: IPNI_HTTP_PROVIDER.md
Comment (BigLep) [reply IPNI_HTTP_PROVIDER.md:143]: Oh I learned about https://github.com/ipni/go-libipni/issues/42 at the end of the doc.  This would be another reason to use reference links.  I would have learned about it right away.

**Review by BigLep — COMMENTED** [review]
Don't block on anything from me.  Just left comments while reading through after it came up during the Content Routing WG on 2023-05-30.  Thanks.

File: IPNI_HTTP_PROVIDER.md
Comment (MarcoPolo) [inline IPNI_HTTP_PROVIDER.md:145]: I've left a comment here: https://github.com/ipni/go-libipni/issues/42#issuecomment-1569330073. More context there, but I don't think we should be putting application logic in the multiaddr.

File: IPNI_HTTP_PROVIDER.md
Comment (lidel) [inline IPNI_HTTP_PROVIDER.md:133]: :thought_balloon:  Is the long term plan  to expose `/cid` and `/multihash`  to `/ipni/v1` namespace too?

File: IPNI_HTTP_PROVIDER.md
Comment (lidel) [inline IPNI_HTTP_PROVIDER.md:151]: I assume this either has a limited TTL where a stale response can be cached for some time,  or should not be cached at all, so the client gets up-to-date response.

Either way, you should require implementer to always return explicit `Cache-Control`:

- To disable HTTP caching and always fetch the latest response:
  ```
  Cache-Control: no-cache, no-store, must-revalidate
  ```
- To allow accepting a cached response and revalidating it in the background, you can use the following Cache-Control header:
   ```
   Cache-Control: max-age=60, stale-while-revalidate=120
   ```
   if the cached response is stale (older than 60 seconds), the client can still serve it while simultaneously sending a request to the server to get updated one -- this removes spikes in latency when Cloudfront  experiences cache miss due to cache age

**Review by lidel — COMMENTED** [review]
@masih added my usual comments about including explicit headers for HTTP Caching + some questions/concerns around `/httpath`

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:143]: > for endpoint of HTTP-only(?) service,

The service may be using vanila TCP or libp2p for transport.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:133]: This is a provider spec only. the `cid` and `multihash` endpoints are for looking up records which reside on the server side and are outside of the scope of this doc.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:145]: "application" is a loose term. From my perspective this is not an application logic in the same way that in a web 2 world URL to a REST API is not an application logic. It's a standard configuration for HTTP API interaction.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:272]: Pagination is achieved by IPLD links across entry chunks. I'll add a paragrah to spell this out. Thanks Steve.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md]: Removing support for `httpath` in multiaddr as per Slack discussion.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:218]: Added notes on caching across the doc. Thanks @lidel

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:233]: It is not clear whether:
1. The entries are included in the response to a request for an advertisement CID
2. One separate request for all the entries is required (unless max content length exceeded)
3. A separate request is required for each block of entries.

Option 2 is the best choice IMO.  Not 1, because a removal advertisement later in the chain may mean that the indexer does not need to download the entries. Also, a utility inspecting the chain may not want the entries. Not 3, because the provider should be able to stream these to the indexer (nd-json encoded).

It should also be stated explicitly that the same endpoint is used to request advertisements as well as entries blocks.

Suggest:
> Entries must be retrieved using a separate request for the CID of the next entries block to fetch. The response contains one or more blocks of entries, up to the maximum content length. The first block in the response must match the CID requested (see entries verification). If more entries remain to be fetched, this is indicated by the last included entries block having a non-empty "Next" field. A subsequent request for the CID specified in the "Next" field retrieves the next set of entries blocks. This continues until the last block has an empty "Next" field.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:301]: Need to specify exactly how to verify an entry:

The CID of each entry chunk can be verified by creating a `schema.EntryChunk` with its `Entries` set to the array of multihashes from the response chunk to verify and the `Next` field set to the CID of the response chunk's `"Next"` field. Then use the IPLD `LinkSystem.ComputeLink` to compute the CID for the `schema.EntryChunk`.

Since the advertisement signature covers the CID of the first entry chunk, verifying the CIDs of the entry chunks also verifies the signature of all the chunks. This works because changing a CID in a chunk alters its CID, and altering the CID, and thus the link in the previous chunk, will change the CID of the previous chunk, all the way up to the first chunk that is covered by the advertisement signature.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:280]: Add "Next" field.
```suggestion
  ],
  "Next": {
    "/": "baguqeeraog3uqu7au2mctrfslzvua5lzl3wyoquqhbyoig6hnputypej5guq"
  }
```

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:316]: Bulk fetch may be harmful when there are context ID removals. If a removal advertisement removes a context ID of an earlier advertisement, then there is no need to download the entries data for that advertisement. However it may still be necessary to download the advertisement to get the signature and link to the previous advertisement.

There should be some mechanism in a bulk fetch that allows only the optional retrieval of the advertisement portion of the data.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:294]: Each chunk can be whatever size the publisher of that data decides, as long as that size and any encoding overhead is below the maximum content length. Once a chunk is created, it cannot be resized since the CID of that chunk, as represented in the advertisement or previous chunks, must be verifiable.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:233]: I think that blocks should be capable of being streamed, using nd-json encoding. This would remove the need to paginate or to create hold a very large response body in memory.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:294]: If we want to have flexible chunk sizes while still being able to validate the data with the provider's signature, then the advertisement signature cannot include the first chunk CID and a separate signature is needed instead.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:233]: * Encoding of body is a separate issue and HTTP offers enough extensibility with Conent-Type and Accept headers to address it. 
* Entries are retrieved by their CID just like an advertisement via `/ipni/v1/ad/{CID}`. The word `ad` in the URI does not imply that it only returns IPLD nodes of type `Advertisement`. Instead it means the endpoint is dedicated to serving ad related content. This is stated [here](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R220).
* I don't think we want to have a separate URI for each IPNI model because to get a CID that one does not know the type of one would have to try all endpoints. It would also defeat the purpose of CIDs a bit?
* This would also solve the pagination problem in the same way that it is currently solved by `Next` link in entries traversal.

Does that make sense? 
In the meantime, I will refine the test to state that the endpoint returns all ad related content if [this](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R220) does not suffice.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:301]: By verification here we simply mean hash digest verification, in that the hash of content corresponds to the CID asked. Anything deeper goes into "ingest" territory IMHO which is out of the scope of this spec, since the spec is focused on the provider side.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:316]: This seems out of the scope of the spec.

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:294]: Is there a specific requirement that we should satisfy which breaks the existing protocol? If not my sense is to push this out of the scope and leave this to be as is: i.e. recommendation on "consider max body sizes when chunking".

Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:298]: The question is, do we want to allow the chunk size, and thus the chunk CID, to change without affecting the advertisement signature.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:234]: > Encoding of body is a separate issue

OK, the specifics of what is put into the body can go into a separate (OpenAPI) spec. 

> Entries are retrieved by their CID just like an advertisement via /ipni/v1/ad/{CID}

Thanks. Just wanted to clarify that.

> I don't think we want to have a separate URI for each IPNI model because to get a CID that one does not know the type of one would have to try all endpoints. It would also defeat the purpose of CIDs a bit?

The point of having a separate URI is to explicitly identify what type of CID is being asked for. If using the ad/ endpoint then the provider only needs to look in the place where ads are kept, and if using an entries/ endpoint then only look where the entries data is kept.  I cannot think of a time when the requestor will not know whether they are requesting an ad or an entry block.  

Anyway, this is OK to use the same endpoint for both. It just means that the provider will not know what kind of object the CID is retrieving, which is find. The provider just has to retrieve it.

 > This would also solve the pagination problem in the same way that it is currently solved by Next link in entries traversal.
 
 The question is about whether there is, or should be, any expectation about 1 entry block per request or about n entry blocks per request.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:305]: This should really be a question about **Signature Validation**.  We should be explicit about whether the advertisement signature covers the CID of the first entry block. I think the answer is yes, but it means that the entries cannot change size without changing the signature on the ads, which may be worth pointing out.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [reply IPNI_HTTP_PROVIDER.md:320]: What that mechanism is, is out of scope. I fo think it is fundamental to advertisement retrieval, bulk or not, that fetching the entries must be optional, so thought it might be worth stating that. Happy to omit it if you think better to leave it out.

Comment (masih) [reply IPNI_HTTP_PROVIDER.md:320]: I don't follow. Can you help me understand what you are asking to be changed here specifically?

File: IPNI_HTTP_PROVIDER.md
Comment (masih) [reply IPNI_HTTP_PROVIDER.md:305]: Singature validation is covered [here](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R299). 

> Should we be explicit about whether the advertisement signature covers the CID of the first entry block?

It currently does and no where in the document we imply otherwise. What's the issue?

> Do we want to allow the chunk size, and thus the chunk CID, to change without affecting the advertisement signature?

Maybe in the future? Right now this seems out of the scope for this spec.

File: IPNI_HTTP_PROVIDER.md
Comment (gammazero) [inline IPNI_HTTP_PROVIDER.md:150]: > topic under which the advertisement is published

Should this say "topic under which the advertisement is announced"?  If this is saying "published" then that seems to imply that the topic would be used to construct the URL path to retrieve the advertisement, but that is not stated anywhere.

**Review by gammazero — APPROVED** [review]

**Pull request merged**
